File: ProgressNotification/app/src/main/java/com/akexorcist/progressnotification/data/AccountNetworkManager.java
package com.akexorcist.progressnotification.data;

import android.os.Handler;

import java.util.concurrent.TimeUnit;

/**
 * Created by Akexorcist on 8/25/2017 AD.
 */

public class AccountNetworkManager {
    public static void applyPremiumAccount(final String userId, final ServiceCallback callback) {
        // Just mock the response
        new Handler().postDelayed(() -> {
                    if (((int) ((Math.random() * 10000)) % 2) != 0) {
                        if (callback != null) {
                            ApplyPremiumAccountResult result = new ApplyPremiumAccountResult(true);
                            callback.onApplyPremiumAccountResult(result);
                        }
                    } else {
                        if (callback != null) {
                            ApplyPremiumAccountException exception = new ApplyPremiumAccountException(userId);
                            callback.onApplyPremiumAccountException(exception);
                        }
                    }
                }, TimeUnit.SECONDS.toMillis(5)
        );

        if (callback != null) {
            callback.onApplyPremiumAccountStarted();
        }
    }

    public interface ServiceCallback {
        void onApplyPremiumAccountStarted();

        void onApplyPremiumAccountResult(ApplyPremiumAccountResult result);

        void onApplyPremiumAccountException(ApplyPremiumAccountException exception);
    }
}

File: app/src/main/java/com/akexorcist/progressnotification/data/AccountNetworkManager.kt
package com.akexorcist.progressnotification.data

import android.os.Handler

import java.util.concurrent.TimeUnit

/**
 * Created by Akexorcist on 8/25/2017 AD.
 */

object AccountNetworkManager {
    fun applyPremiumAccount(userId: String, callback: ServiceCallback?) {
        // Just mock the response
        Handler().postDelayed({
            if ((Math.random() * 10000).toInt() % 2 != 0) {
                callback?.onApplyPremiumAccountResult(ApplyPremiumAccountResult(true))
            } else {
                callback?.onApplyPremiumAccountException(ApplyPremiumAccountException(userId))
            }
        }, TimeUnit.SECONDS.toMillis(5))
        callback?.onApplyPremiumAccountStarted()
    }

    interface ServiceCallback {
        fun onApplyPremiumAccountStarted()

        fun onApplyPremiumAccountResult(result: ApplyPremiumAccountResult)

        fun onApplyPremiumAccountException(exception: ApplyPremiumAccountException)
    }
}
